src/sinks/util/sink.rs

```diff
@@ -424,163 +310,137 @@ where
-                                self.poll_complete()?;
-                            }
-                            return Ok(AsyncSink::Ready);
-                        }
-                        PushResult::Overflow(item) => {
-                            match self.handle_full_batch(item, &partition) {
-                                FullBatchResult::Result(result) => return result,
-                                FullBatchResult::Continue(item) => item,
-                            }
-                        }
-                    }
-                }
+        let batch = loop {
+            if let Some(batch) = self.partitions.get_mut(&partition) {
+                break batch;
             }
-            None => item,
-        };
 
-        trace!("Replacing batch.");
-        // We fall through to this case, when there is no batch already
-        // or the batch got submitted by polling_complete above.
-        let mut batch = self.batch.fresh();
+            let batch = self.batch.fresh();
+            self.partitions.insert(partition.clone(), batch);
 
-        match batch.push(item) {
-            PushResult::Overflow(_) => unreachable!("Empty buffer overflowed"),
-            PushResult::Ok(full) => {
-                self.set_linger(partition.clone());
+            let delay = delay_for(self.timeout);
+            self.lingers.insert(partition.clone(), delay);
+        };
 
-                self.partitions.insert(partition, batch);
+        if let PushResult::Overflow(item) = batch.push(item) {
+            self.buffer = Some((partition, item));
+        }
 
-                if full {
-                    self.poll_complete()?;
-                }
+        Ok(())
+    }
 
-                Ok(AsyncSink::Ready)
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        loop {
+            // Poll inner service while not ready, if we don't have buffer or any batch.
+            if self.buffer.is_none() && self.partitions.is_empty() {
+                ready!(self.service.poll_complete(cx));
+                return Poll::Ready(Ok(()));
             }
-        }
-    }
 
-    fn poll_complete(&mut self) -> Poll<(), Self::SinkError> {
-        self.service.poll_complete()?;
+            // Try send batches.
+            let this = self.as_mut().project();
+            let mut partitions_ready = vec![];
+            for (partition, batch) in this.partitions.iter() {
+                let should_send = {
+                    if (*this.closing && !batch.is_empty()) || batch.was_full() {
+                        true
+                    } else {
+                        let linger = this
+                            .lingers
+                            .get_mut(&partition)
+                            .expect("linger should exists for poll_flush");
+                        matches!(linger.poll_unpin(cx), Poll::Ready(()))
+                    }
+                };
```

Comment:
```suggestion
                if (*this.closing && !batch.is_empty())
                    || batch.was_full()
                    || matches!(
                        this
                            .lingers
                            .get_mut(&partition)
                            .expect("linger should exists for poll_flush")
                            .poll_unpin(cx),
                        Poll::Ready(())
                       ) {
```
(and drop the `if should_send {` below)

You might even be able to turn this into a pure iterator:
```rust
    let partitions_ready: Vec<_> = this.partitions.iter()
      .filter(|(partition, batch)| …above condition…)
      .map(|(partition, _)| partition.clone())
      .collect();
```

Reply:
iter-filter-map-collect will not work due to borrow checker errors :disappointed: but I'll change `should_send`